### DISCUSSION TITLE
Separate files for `view!`

### DISCUSSION THREAD

**Discussion by Grafcube** [leptos-rs/leptos#865] [Ideas]
Is there any way I can write the html in external files and then import them into view? I'm using neovim and I would like to use the existing html and tailwind language servers for formatting, documentation, autocomplete etc.

**Comment by gbj** [comment]
Well, it's not HTML, is the thing -- it's a Rust proc macro. In other words, `view` doesn't take a string, it takes a stream of Rust tokens as its argument. I guess if you wanted the Angular-style* "HTML in one file, CSS in another, Rust in another" you could write another macro that includes a string from a file, parses that into a token stream with `syn`, and then feeds that into `view`. you'd lose the ability to have any kind of rust-analyzer support for expressions inside the view macro though (for example, LSP hover support on the component name should give you component docs, etc.)

**Reply by Grafcube** [reply]
Well, mainly I would just like to get the formatting+automatic tag closing and tailwind suggestions directly from my lsp, with the hover doc showing the exact css that the class applies. Unfortunately this doesn't work in rust files. Additionally the syntax highlighting isn't very accurate with strings. While I can manage without these things, it's still tedious without tailwind completions. I have to manually import tags anyway since code actions and auto-import doesn't seem to work with `view`, so I figured I could at least leverage the html and tailwind lsp even if it chokes on some differences. It seems to handle it just fine anyway without any warnings.

I don't really know how to write a macro that could read an external file. I tried something with `include_str` but that obviously didn't work. Ideally I would suggest something like `.rsx` files which have support for both the usual lsp tools as well as being able to import Rust types but I have no idea how something like that could even be implemented. For now, I think a way to use tailwind's lsp, either in a separate file or within rust, would greatly improve the experience for me.

**Reply by mondeja** [reply]
I suspect that the easiest way to solve this problem would be to create extractors/matchers for intellisense extensions as plugins. Currently with VSCode +  TailwindCSS Intellisense extension if you configure `tailwindCSS.includeLanguages.rust: "javascript"` you'll get Tailwind autocompletion for `class` attributes in literal strings, but not if strings are inside closures.

**Reply by Grafcube** [reply]
I use neovim so it's a little different. I managed to get Tailwind working by following [this thread](https://old.reddit.com/r/neovim/comments/yukgxy/rust_yew_tailwindcss_intellisense/). Essentially, all I had to do was add `"rust"` to the `filetypes` table and set `lsp.tailwindcss.init_options.userLanguages.rust = "html"`. The colorizer plugin worked fine with tailwind set to `"lsp"`.

![image](https://user-images.githubusercontent.com/34889224/232059131-d29bfeb9-107c-4884-8ef2-d3315a298f07.png)

At this point the only issues left are minor annoyances that I can easily ignore but would like to see addressed at some point. A formatter would be convenient, as would automatically closing tags. I have no idea how this could be done but I vaguely remember reading that it's possible to change rust analyzer's behaviour in macros. As for syntax highlighting, on neovim at least, it's possible to use injections with treesitter.

**Reply by mondeja** [reply]
For formatting in VSCode I'm currently using leptosfmt with the RunOnSave extension, the following setting does the trick:

```json
"emeraldwalk.runonsave": {
  "commands": [
    {
      "match": "\\.rs",
      "isAsync": true,
      "cmd": "leptosfmt ${file}"
    }
  ]
}
```

**Reply by Grafcube** [reply]
I didn't know that existed. I've installed it and made an autocommand for format on save.

```lua
vim.api.nvim_create_autocmd(
  { "BufWritePost" },
  {
    pattern = { "*.rs" },
    desc = "Leptos Formatter",
    callback = function(_)
      local buf = vim.fn.expand("%:p")
      local sw = vim.api.nvim_buf_get_option(0, "shiftwidth")
      os.execute("leptosfmt -t " .. sw .. " " .. buf)
      vim.cmd("checktime")
    end
  }
)
```

The syntax highlighting is a little weird though.

![image](https://user-images.githubusercontent.com/34889224/232200260-1fccec74-a880-45e6-a5f9-367082dbdd00.png)

Also, rust analyzer seems to be a lot slower to start up after I set up my Cargo.toml for leptos but I'm not sure if that's related.

**Reply by gbj** [reply]
To me, this seems like a kind of weird treesitter issue, honestly. (I think nvim uses treesitter for highlighting right?) rust-analyzer emits syntactic highlighting info that works correctly, but it seems like nvim doesn't use it, maybe?
